Source/scrollrt.cpp

```diff
@@ -916,6 +917,139 @@ void DrawDungeon(const Surface &out, Point tilePosition, Point targetBufferPosit
+	    || (tilePosition.y > 95 && IsAnyOf(tilePosition.x, 32, 84))) {
+		// left river bank
+		pos = { 68, 22 - abs(tilePosition.y % 3) };
+	} else if ((tilePosition.y < 0 && IsAnyOf(tilePosition.x, 3, 63, 83))
+	    || (tilePosition.y > 95 && IsAnyOf(tilePosition.x, 33, 85))) {
+		// right river bank
+		pos = { 69, 22 - abs(tilePosition.y % 3) };
+	}
+	return pos;
+}
+
+Point FillPattern(Point tilePosition)
+{
+	constexpr int PATTERNWIDTH = 5;
+	constexpr int PATTERNHEIGHT = 5;
+	constexpr std::array<Point, 6> patterns { { { 51, 27 }, { 23, 57 }, { 54, 34 }, { 79, 46 }, { 41, 32 }, { 52, 21 } } };
+
+	static int patternRandomizer[MAXDUNX][MAXDUNY] {};
+	static bool doInit = true;
+
+	if (doInit) {
+		doInit = false;
+		std::mt19937 BetterRng;
+		BetterRng.seed(666);
+		std::uniform_int_distribution<int> dist(0, patterns.size() - 1);
+		for (int x = 0; x < MAXDUNX; x++) {
+			for (int y = 0; y < MAXDUNY; y++) {
+				patternRandomizer[x][y] = dist(BetterRng);
+			}
+		}
+	}
+
+	Point patternBaseTile = abs(tilePosition);
+	int offsetX = patternBaseTile.x % PATTERNWIDTH;
+	int offsetY = patternBaseTile.y % PATTERNHEIGHT;
+	if (tilePosition.x < 0)
+		offsetX = PATTERNWIDTH - 1 - patternBaseTile.x % PATTERNWIDTH;
+	if (tilePosition.y < 0)
+		offsetY = PATTERNHEIGHT - 1 - patternBaseTile.y % PATTERNHEIGHT;
+	Displacement patternOffset { offsetX, offsetY };
+
+	patternBaseTile.x /= PATTERNWIDTH;
+	patternBaseTile.y /= PATTERNHEIGHT;
+
+	Point patternPos = CheckPredefinedPattern(tilePosition);
+	if (patternPos != Point { 0, 0 }) {
+		if (InDungeonBounds(patternBaseTile)) {
+			patternRandomizer[patternBaseTile.x][patternBaseTile.y] = -1;
+		}
+		return patternPos;
+	}
+
+	Point base = patterns[0];
+	if (InDungeonBounds(patternBaseTile)) {
+		int pattern = patternRandomizer[patternBaseTile.x][patternBaseTile.y];
+		if (pattern != -1) {
+			base = patterns[pattern];
+		}
+	}
+	return base + patternOffset;
+}
+
+void DrawDungeonBorders(const Surface &out, Point tilePosition, Point targetBufferPosition)
+{
+	assert(InDungeonBounds(tilePosition));
+
+	LightTableIndex = dLight[tilePosition.x][tilePosition.y];
+
+	DrawCell(out, tilePosition, targetBufferPosition);
+
+	// Tree leaves should always cover player when entering or leaving the tile,
+	// So delay the rendering until after the next row is being drawn.
+	// This could probably have been better solved by sprites in screen space.
+	if (tilePosition.x > 0 && tilePosition.y > 0 && targetBufferPosition.y > TILE_HEIGHT) {
+		char bArch = dSpecial[tilePosition.x - 1][tilePosition.y - 1];
+		if (bArch != 0) {
+			CelDrawTo(out, targetBufferPosition + Displacement { 0, -TILE_HEIGHT }, *pSpecialCels, bArch);
+		}
+	}
+}
+
+/**
+ * @brief Render some semi random patterns beyond town borders instead of black void
+ */
+
+void RenderTownBorders(const Surface &out, Point tilePosition, Point targetBufferPosition, bool isDungeon)
+{
+	if (leveltype != DTYPE_TOWN)
+		return;

```

Comment:
This blocks the cases where `isDungeon == true`

Reply:
no